utils/retrieve-state/src/lib.rs

```diff
@@ -289,50 +337,369 @@ pub fn put_block_with_deploys(
+        .flatten()
+        .map(|peer: Peer| (peer.address, peer))
+        .collect::<HashMap<SocketAddr, Peer>>();
 
-    let mut start = Instant::now();
-    let mut tries_downloaded = 0;
-    while let Some(next_trie_key) = outstanding_tries.pop() {
-        let read_result = get_trie(
-            client,
-            url,
-            GetTrieParams {
-                trie_key: next_trie_key,
-            },
-        )
-        .await?;
-        if let Some(blob) = read_result.maybe_trie_bytes {
-            let bytes: Vec<u8> = blob.into();
-            let (trie, _): (Trie<Key, StoredValue>, _) = FromBytes::from_bytes(&bytes)?;
-            let mut missing_descendants = engine_state
-                .put_trie_and_find_missing_descendant_trie_keys(CorrelationId::new(), &trie)?;
-            outstanding_tries.append(&mut missing_descendants);
-            tries_downloaded += 1;
-            progress_fn()
-        } else {
-            return Err(anyhow::anyhow!(
-                "unable to download trie at {:?}",
-                next_trie_key
-            ));
+    println!("got {}/{} valid peers", peer_map.len(), peers.len());
+
+    let mut total_tries_count = 0;
+
+    let mut outstanding_trie_keys = vec![state_root_hash];
+    let mut tries_downloaded = 0usize;
+    let mut in_flight_counters = peers
+        .iter()
+        .map(|addr| (addr, 0usize))
+        .collect::<HashMap<_, _>>();
+
+    loop {
+        // Top level task keeps track of missing descendants and work distribution.
+        for (address, peer) in peer_map.iter() {
+            let counter = in_flight_counters.get_mut(&address).unwrap();
+            // TODO: understand what's happening when requests get backed up, causing starvation of
+            // base_recv
+            if *counter < concurrent_requests_per_peer {
+                if let Some(next_trie_key) = outstanding_trie_keys.pop() {
+                    let mut sender = peer.sender.clone();
+                    *counter += 1;
+                    sender.send(PeerMsg::GetTrie(next_trie_key)).await.unwrap();
+                } else {
+                    break;
+                }
+            }
+        }
+
+        if let Some(msg) = base_recv.next().await {
+            match msg {
+                Ok(PeerMsg::TrieDownloaded {
+                    trie,
+                    peer,
+                    elapsed: _,
+                }) => {
+                    let engine_state = Arc::clone(&engine_state);
+                    /*
+                    let mut missing_trie_descendants = tokio::task::spawn_blocking(move || {
+                        engine_state
+                            .put_trie_and_find_missing_descendant_trie_keys(
+                                CorrelationId::new(),
+                                &trie,
+                            )
+                            .unwrap()
+                    })
+                    .await
+                    .unwrap();
+                    */
```

Comment:
Did you want to leave this comment in?

Reply:
I was trying out spawn_blocking, but this actually led to some performance issues where we overuse the threadpool. Removed.